### PR TITLE
DAG block synchronization

### PR DESCRIPTION
**Pull request by lochjin**

### CODE REVIEW

File: core/message/msgsyncdag.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+type MsgSyncDAG struct {
+	MainLocator []*hash.Hash
+	GS          *blockdag.GraphState
+}
+
+func (msg *MsgSyncDAG) Decode(r io.Reader, pver uint32) error {
+	var count uint64
+	err := s.ReadElements(r, &count)
+	if err != nil {
+		fmt.Println(err)
```
Comment (dindinw) [inline core/message/msgsyncdag.go]: unmanaged debug msg

File: core/message/msgsyncdag.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+	msg.GS = blockdag.NewGraphState()
+	err = msg.GS.Decode(r, pver)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (msg *MsgSyncDAG) Encode(w io.Writer, pver uint32) error {
+	fmt.Println(uint64(len(msg.MainLocator)))
```
Comment (dindinw) [inline core/message/msgsyncdag.go]: unmanaged debug msg

**Comment by lochjin** [comment]
> The initial downloading cannot finish after applying this pull request.
> 
> https://drive.google.com/file/d/1fzMFjb0jrOowGo7NXCD-JKiKlW0SHGwl/view?usp=sharing

This problem is a performance problem. You may need to clear the data. I will optimize it later.

**Comment by forchain** [comment]
> performance

OK, I will clear the data and try again

**Comment by lochjin** [comment]
> > performance
> 
> OK, I will clear the data and try again

https://github.com/Qitmeer/qitmeer/pull/202/commits/c61d00852d730547c5ae6687dcbf1c961e4fde14  This problem has been fixed

**Comment by dindinw** [comment]
already merged to https://github.com/Qitmeer/qitmeer/tree/0.8.4-dev

**Pull request closed without merging**